client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,341 @@
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::{result, collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
+use std::time::Instant;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber, Level, span::{Attributes, Record, Id}};
+use tracing_subscriber::CurrentSpan;
+
+use sc_client_api::BlockBackend;
+use sc_rpc_server::MAX_PAYLOAD;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Heuristic for average event size in bytes.
+const AVG_EVENT: usize = 600 * 8;
+// Heuristic for average span size in bytes.
+const AVG_SPAN: usize = 100 * 8;
+// Estimate of the max base RPC payload size when the Id is bound as a u64. If strings
+// are used for the RPC Id this may need to be adjusted. Note: The base payload
+// does not include the RPC result.
+//
+// The estimate is based on the JSONRPC response message which has the following format:
+// `{"jsonrpc":"2.0","result":[],"id":18446744073709551615}`.
+//
+// We care about the total size of the payload because jsonrpc-server will simply ignore
+// messages larger than `sc_rpc_server::MAX_PAYLOAD` and the caller will not get any
+// response.
+const BASE_PAYLOAD: usize = 100;
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &str = "pallet,frame,state";
+const TRACE_TARGET: &str = "block_trace";
+// Default to not filtering based on storage keys as storage keys vary per chain.
+const DEFAULT_STORAGE_KEYS: &str = "";
+// The name of a field required for all events.
+const REQUIRED_EVENT_FIELD: &str  = "method";
+
+/// Tracing Block Result type alias
+pub type Result<T> = result::Result<T, Error>;
+
+/// Tracing Block error
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+#[non_exhaustive]
+pub enum Error {
+	#[error("Invalid block Id: {0}")]
+	InvalidBlockId(#[from] sp_blockchain::Error),
+	#[error("Missing block component: {0}")]
+	MissingBlockComponent(String),
+	#[error("Dispatch error: {0}")]
+	Dispatch(String)
+}
```

Comment:
@dvdplm or @niklasad1 Does this look ok for errors? (As opposed to the plain `String` errors we used to have)

Reply:
Works.

I changed the name of the `Result` type alias to `TraceBlockResult` for clarity.